File: 0652-find-duplicate-subtrees/0652-find-duplicate-subtrees.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    unordered_map<string, int> ump;
    vector<TreeNode*> ans;
    string serialize(TreeNode* root) {
        if (!root) return "#";

        string curr_subtree = to_string(root->val) + "," + serialize(root->left) + "," + serialize(root->right);

        if (ump.count(curr_subtree)) {
            cout<<curr_subtree<<endl;
            if (ump[curr_subtree] < 2) ans.push_back(root);
        }
        ump[curr_subtree]++;

        return curr_subtree;
    }
    vector<TreeNode*> findDuplicateSubtrees(TreeNode* root) {
        string dummy = serialize(root);
        return ans;
    }
};

File: 0652-find-duplicate-subtrees/0652-find-duplicate-subtrees.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
        

    def findDuplicateSubtrees(self, root: Optional[TreeNode]) -> List[Optional[TreeNode]]:
        seen = collections.defaultdict(int)
        ans = []
        
        def serialize(root):
            if root is None:
                return "#"

            curr_subtree = str(root.val) + "," + serialize(root.left) + "," + serialize(root.right)

            if curr_subtree in seen:
                if seen[curr_subtree] < 2:
                    print(curr_subtree)
                    ans.append(root)
            
            seen[curr_subtree] += 1

            return curr_subtree

        serialize(root)
        return ans
        
        